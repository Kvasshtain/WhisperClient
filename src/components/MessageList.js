import React from 'react'
import PropTypes from 'prop-types'

class MessageList extends React.Component {
    
    renderMessageList = () => {
        const {messages} = this.props;

        if (messages.length) {
            return messages.map(function (item, index) {
                return (
                    <p key = {index}>{item.messages}</p>
                )
            })
        }
    }

    render() {
        return (
            <div className="MessageList">
                {this.renderMessageList()}
            </div>
        )
    }
}

MessageList.propTypes = {
    messages: PropTypes.arrayOf(PropTypes.string.isRequired)
}

export {MessageList}