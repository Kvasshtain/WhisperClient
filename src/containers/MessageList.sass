@import "ScrollList.sass"

$no-border: 0px solid white

.message-list
    @include scroll-list
    > div
        margin: 10px
        border-radius: 10px
        background-color: #d7f1dd
        border: $no-border
        padding: 10px

$no-border: 0px solid white
$font-color: #233527
$scroll-down-button-font-size: 50px
$scroll-down-button-size: 75px
$scroll-down-button-margin: 65px

.scroll-down-button
    position: absolute
    display: flex
    right: 0
    bottom: 0
    margin: $scroll-down-button-margin
    align-items: center
    justify-content: center
    width: $scroll-down-button-size
    height: $scroll-down-button-size
    border-radius: $scroll-down-button-size
    background-color: #8bcc9a
    border: $no-border
    font-size: $scroll-down-button-font-size
    color: $font-color
    opacity: 0.3

.scroll-down-button:hover
    opacity: 1
