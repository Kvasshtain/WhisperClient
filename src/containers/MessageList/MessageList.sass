@import "../ScrollList.sass"

$no-border: 0px solid white

.message-list
    @include scroll-list
    > div
        display: flex
        flex-direction: column
        justify-content: flex-start
        > div.current_user_message
            align-self: flex-end
        > div.another_user_message
            align-self: flex-start