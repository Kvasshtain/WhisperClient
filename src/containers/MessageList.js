import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import {
  fetchMessagesList,
  fetchNewMessages,
  subscribeForNewMessages,
  unsubscribeForNewMessages,
} from '../actions/messageListActions'

import { MessageFrame } from '../components/MessageFrame/MessageFrame'
import { NewMessageInput } from '../components/NewMessageInput'
//import { updateInterval } from '../applicationSettings'
import './MessageList.sass'

class MessageList extends React.Component {
  constructor(props) {
    super(props)

    this.messageListRef = React.createRef()

    this.state = {
      enableScrollDown: true,
      previousMessagesLength: 0,
      suspendMessagesFetching: false,
      currentChat: null,
    }
  }

  componentDidMount = () => {
    const { currentChat, subscribeForNewMessages } = this.props
    if (this.state.currentChat !== currentChat) {
      this.setState({ currentChat })
      subscribeForNewMessages(this.props.currentChat._id)
    }
    this.fetchMessages()
    //this.initializeMessagesListUpdateTimer()
  }

  // initializeMessagesListUpdateTimer = () => {
  //   this.timerID = setInterval(() => {
  //     this.fetchNewMessages()
  //     this.tryLoadMessagesUntilScrollAppears()
  //   }, updateInterval)
  // }

  componentWillUnmount() {
    this.props.unsubscribeForNewMessages()
    //clearInterval(this.timerID)
  }

  componentDidUpdate = () => {
    const { messages, currentChat, subscribeForNewMessages } = this.props
    const scrollDownShift = 30
    const { current } = this.messageListRef

    if (this.state.currentChat !== currentChat) {
      this.reloadMessagesList()
      subscribeForNewMessages(this.props.currentChat._id)
      return
    }

    if (!messages) return

    const messagesLength = messages.length
    const previousMessagesLength = this.state.previousMessagesLength

    if (messagesLength === previousMessagesLength) return

    this.setState({
      previousMessagesLength: messagesLength,
    })

    this.setState({
      suspendMessagesFetching: false,
    })

    if (current) {
      current.scrollTop += scrollDownShift
    }

    this.scrollDownIfEnabled()
  }

  reloadMessagesList = () => {
    this.setState({
      currentChat: this.props.currentChat,
    })

    this.fetchMessagesForced()
    this.setState({
      suspendMessagesFetching: true,
    })
  }

  fetchMessages = () => {
    if (this.state.suspendMessagesFetching) return

    this.fetchMessagesForced()
  }

  fetchMessagesForced = () => {
    const currentChat = this.props.currentChat
    const messages = this.props.messages

    if (!currentChat || !currentChat._id) return
    if (!messages) return

    const time = new Date().getTime()

    const oldestMessageTime = messages.length ? messages[0].time : time

    this.props.fetchMessagesList(currentChat._id, oldestMessageTime)
  }

  fetchNewMessages = () => {
    const currentChat = this.props.currentChat
    const messages = this.props.messages

    if (!currentChat || !currentChat._id) return
    if (!messages) return

    const lastMessageIndex = messages.length - 1

    if (lastMessageIndex < 0) {
      this.fetchMessages()
      return
    }

    this.props.fetchNewMessages(
      currentChat._id,
      messages[lastMessageIndex].time
    )
  }

  scrollDownIfEnabled = () => {
    if (this.state.enableScrollDown) {
      this.setState({
        enableScrollDown: false,
      })

      this.scrollDown()
    }
  }

  tryLoadMessagesUntilScrollAppears() {
    const { current } = this.messageListRef

    if (current.offsetWidth <= current.clientWidth) {
      this.fetchMessages()
    }
  }

  scrollDown = () => {
    const { current } = this.messageListRef

    if (current) {
      current.scrollTop = current.scrollHeight
    }
  }

  renderMessageList = () => {
    const { messages, currentUser } = this.props
    const messagesLength = messages.length

    if (messages && messagesLength) {
      return messages.map(function(item, index) {
        return (
          <div key={index}>
            <MessageFrame message={item} currentUserEmail={currentUser.email} />
          </div>
        )
      })
    }
  }

  onScrollDownClick = () => {
    this.scrollDown()
  }

  onScroll = () => {
    const minScrollTop = 20
    const { current } = this.messageListRef

    if (current.scrollTop < minScrollTop) {
      this.setState({
        suspendMessagesFetching: true,
      })

      this.fetchMessages()
    }
  }

  sendNewMessage = newMessage => {
    this.props.sendNewMessage(newMessage)

    this.setState({
      enableScrollDown: true,
    })
  }

  render() {
    const strToDecode = '&#11015'
    const parser = new DOMParser()
    const decodedString = parser.parseFromString(
      `<!doctype html><body>${strToDecode}`,
      'text/html'
    ).body.textContent

    return (
      <div>
        <div className="scroll-down-button" onClick={this.onScrollDownClick}>
          {decodedString}
        </div>
        <div
          ref={this.messageListRef}
          className="message-list"
          onScroll={this.onScroll}
        >
          {this.renderMessageList()}
        </div>
        <NewMessageInput onSubmitNewMessage={this.sendNewMessage} />
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    messages: state.messages,
    currentChat: state.currentChat,
    currentUser: state.currentUser,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    fetchMessagesList: (chatId, oldestMessageTime) =>
      dispatch(fetchMessagesList(chatId, oldestMessageTime)),
    fetchNewMessages: (chatId, newestMessageTime) =>
      dispatch(fetchNewMessages(chatId, newestMessageTime)),
    subscribeForNewMessages: chatId =>
      dispatch(subscribeForNewMessages(chatId)),
    unsubscribeForNewMessages: () => dispatch(unsubscribeForNewMessages()),
  }
}

MessageList.propTypes = {
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      message: PropTypes.exact({
        chatId: PropTypes.number.isRequired,
        authorName: PropTypes.string.isRequired,
        authorEmail: PropTypes.string.isRequired,
        time: PropTypes.oneOfType([
          PropTypes.number,
          PropTypes.string,
          PropTypes.instanceOf(Date),
        ]).isRequired,
        text: PropTypes.string.isRequired,
        wasMessageReceived: PropTypes.bool,
      }),
    })
  ),
}

export { MessageList }
export default connect(mapStateToProps, mapDispatchToProps)(MessageList)
