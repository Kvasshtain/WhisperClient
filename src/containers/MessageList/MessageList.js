import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import {
  fetchMessagesList,
  fetchNewMessages,
  subscribeForNewMessages,
  unsubscribeForNewMessages,
} from '../../actions/messageListActions'

import { MessageFrame } from '../../components/MessageFrame/MessageFrame'
import { ScrollDownButton } from './__ScrollDownButton/MessageList-ScrollDownButton'

import './MessageList.sass'
import './__Item/MessageList-Item.sass'
import './__ChatNameCaption/MessageList-ChatNameCaption.sass'

class MessageList extends React.Component {
  constructor(props) {
    super(props)

    this.messageListRef = React.createRef()

    this.state = {
      enableScrollDown: true,
      previousMessagesLength: 0,
      suspendMessagesFetching: false,
      currentChat: null,
      needScrollDown: true,
      showScrollDownButton: false,
    }
  }

  componentDidMount = () => {
    const { currentChat, subscribeForNewMessages } = this.props
    if (this.state.currentChat !== currentChat) {
      this.setState({ currentChat })
      subscribeForNewMessages(this.props.currentChat._id)
    }
    this.fetchMessages()
  }

  componentWillUnmount() {
    this.props.unsubscribeForNewMessages()
  }

  componentDidUpdate = () => {
    const { messages, currentChat, subscribeForNewMessages } = this.props
    const scrollDownShift = 30
    const { current } = this.messageListRef

    if (this.state.currentChat !== currentChat) {
      this.reloadMessagesList()
      subscribeForNewMessages(this.props.currentChat._id)
      return
    }

    if (!messages) return

    const messagesLength = messages.length
    const previousMessagesLength = this.state.previousMessagesLength

    if (messagesLength <= previousMessagesLength) {
      return
    }

    this.setState({
      previousMessagesLength: messagesLength,
      suspendMessagesFetching: false,
    })

    if (current) {
      current.scrollTop += scrollDownShift
    }

    this.scrollDownIfNeed()
  }

  reloadMessagesList = () => {
    this.setState({
      currentChat: this.props.currentChat,
      previousMessagesLength: 0,
      suspendMessagesFetching: true,
      needScrollDown: true,
    })

    this.fetchMessagesForced()
  }

  fetchMessages = () => {
    if (this.state.suspendMessagesFetching) return

    this.fetchMessagesForced()
  }

  fetchMessagesForced = () => {
    const currentChat = this.props.currentChat
    const messages = this.props.messages

    if (!currentChat || !currentChat._id) return
    if (!messages) return

    const time = new Date().getTime()

    const oldestMessageTime = messages.length ? messages[0].time : time

    this.props.fetchMessagesList(currentChat._id, oldestMessageTime)
  }

  fetchNewMessages = () => {
    const currentChat = this.props.currentChat
    const messages = this.props.messages

    if (!currentChat || !currentChat._id) return
    if (!messages) return

    const lastMessageIndex = messages.length - 1

    if (lastMessageIndex < 0) {
      this.fetchMessages()
      return
    }

    this.props.fetchNewMessages(
      currentChat._id,
      messages[lastMessageIndex].time
    )
  }

  scrollDownIfNeed = () => {
    if (this.state.needScrollDown) {
      this.setState({
        needScrollDown: false,
      })

      this.scrollDown()
    }
  }

  tryLoadMessagesUntilScrollAppears() {
    const { current } = this.messageListRef

    if (current.offsetWidth <= current.clientWidth) {
      this.fetchMessages()
    }
  }

  scrollDown = () => {
    const { current } = this.messageListRef

    this.setState({
      needScrollDown: false,
    })

    if (current) {
      current.scrollTop = current.scrollHeight
    }
  }

  onScrollDownClick = () => {
    this.scrollDown()
  }

  onScroll = () => {
    const minScrollTop = 200
    const scrollDownButtonThreshold = 200
    const {
      scrollHeight,
      scrollTop,
      clientHeight,
    } = this.messageListRef.current

    this.setState({
      showScrollDownButton:
        scrollHeight - scrollTop >= clientHeight + scrollDownButtonThreshold,
      needScrollDown: scrollHeight - scrollTop === clientHeight,
    })

    if (this.state.suspendMessagesFetching) return

    if (scrollTop < minScrollTop) {
      this.setState({
        suspendMessagesFetching: true,
      })

      this.fetchMessages()
    }
  }

  renderMessageList = (messages, hasServerReceivedMessage) => {
    const { currentUser } = this.props
    const messagesLength = messages.length

    if (messages && messagesLength) {
      return messages.map(function(item, index) {
        return (
          <div className="item" key={index}>
            <MessageFrame hasServerReceivedMessage={hasServerReceivedMessage} message={item} currentUserEmail={currentUser.email} />
          </div>
        )
      })
    }
  }

  renderScrollDownButton = () => {
    if (this.state.showScrollDownButton) {
      return <ScrollDownButton onScrollDownClick={this.onScrollDownClick} />
    }
  }

  renderChatNameCaption = () => {
    const chatName = this.props.currentChat.name
    if (chatName) {
      return (
        <div className="chatNameCaption">
          <div>
            <span>{chatName}</span>
          </div>
        </div>
      )
    }
  }

  render() {
    const strToDecode = '&#11015'
    const parser = new DOMParser()
    const decodedString = parser.parseFromString(
      `<!doctype html><body>${strToDecode}`,
      'text/html'
    ).body.textContent

    const { messages, waitingMessages } = this.props

    return (
      <React.Fragment>
        <div
          ref={this.messageListRef}
          className="messageList"
          onScroll={this.onScroll}
        >
          {this.renderMessageList(messages, true)}
          {this.renderMessageList(waitingMessages, false)}
        </div>
        {this.renderScrollDownButton()}
        {this.renderChatNameCaption()}
      </React.Fragment>
    )
  }
}

const mapStateToProps = state => {
  return {
    messages: state.messages,
    waitingMessages: state.messagesWaitingList,
    currentChat: state.currentChat,
    currentUser: state.currentUser,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    fetchMessagesList: (chatId, oldestMessageTime) =>
      dispatch(fetchMessagesList(chatId, oldestMessageTime)),
    fetchNewMessages: (chatId, newestMessageTime) =>
      dispatch(fetchNewMessages(chatId, newestMessageTime)),
    subscribeForNewMessages: chatId =>
      dispatch(subscribeForNewMessages(chatId)),
    unsubscribeForNewMessages: () => dispatch(unsubscribeForNewMessages()),
  }
}

MessageList.propTypes = {
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      message: PropTypes.exact({
        chatId: PropTypes.number.isRequired,
        authorName: PropTypes.string.isRequired,
        authorEmail: PropTypes.string.isRequired,
        time: PropTypes.oneOfType([
          PropTypes.number,
          PropTypes.string,
          PropTypes.instanceOf(Date),
        ]).isRequired,
        text: PropTypes.string.isRequired,
        wasMessageReceived: PropTypes.bool,
      }),
    })
  ),
}

export { MessageList }
export default connect(mapStateToProps, mapDispatchToProps)(MessageList)
