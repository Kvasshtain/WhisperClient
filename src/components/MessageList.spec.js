import React from 'react'
import {shallow} from 'enzyme'
import { shallowToJson } from 'enzyme-to-json';

import {MessageList} from './MessageList'

describe('MessageList', () => {
    const mockMessages = ['test message']
    
    it ('should render correctly', () => {
        const myComponent = shallow(<MessageList 
            messages = {mockMessages}
        />)

        expect(shallowToJson(myComponent)).toMatchSnapshot();
      })
    
  })
