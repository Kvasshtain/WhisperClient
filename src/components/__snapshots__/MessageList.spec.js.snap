// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should render correctly 1`] = `
<div
  className="MessageList"
>
  <p
    key="0"
  />
</div>
`;
